Source/engine/demomode.cpp

```diff
@@ -0,0 +1,312 @@
+/**
+ * @file animationinfo.h
+ *
+ * Contains most of the the demomode specific logic
+ */
+
+#include <deque>
+#include <fstream>
+#include <sstream>
+#include <iostream>
+
+#include "demomode.h"
+#include "utils/display.h"
+#include "utils/paths.h"
+#include "menu.h"
+#include "options.h"
+#include "nthread.h"
+#include "pfile.h"
+
+namespace devilution {
+
+namespace {
+
+enum class DemoMsgType {
+	GameTick = 0,
+	Rendering = 1,
+	Message = 2,
+};
+
+struct demoMsg {
+	DemoMsgType type;
+	uint32_t message;
+	int32_t wParam;
+	int32_t lParam;
+	float progressToNextGameTick;
+};
+
+std::ofstream demoRecording;
+std::deque<demoMsg> demo_message_queue;
+uint32_t demoModeLastTick = 0;
+
+int logicTick = 0;
+int startTime = 0;
+
+int demoGraphicsWidth = 640;
+int demoGraphicsHeight = 480;
+
+void PumpDemoMessage(DemoMsgType demoMsgType, uint32_t message, int32_t wParam, int32_t lParam, float progressToNextGameTick)
+{
+	demoMsg msg;
+	msg.type = demoMsgType;
+	msg.message = message;
+	msg.wParam = wParam;
+	msg.lParam = lParam;
+	msg.progressToNextGameTick = progressToNextGameTick;
+
+	demo_message_queue.push_back(msg);
+}
+
+bool LoadDemoMessages(int i)
+{
+	std::ifstream demofile;
+	char demoFilename[16];
+	snprintf(demoFilename, 15, "demo_%d.dmo", i);
+	demofile.open(paths::PrefPath() + demoFilename);
+	if (!demofile.is_open()) {
+		return false;
+	}
+
+	std::string line;
+	std::getline(demofile, line);
+	std::stringstream header(line);
+
+	std::string number;
+	std::getline(header, number, ','); // Demo version
+	if (std::stoi(number) != 0) {
+		return false;
+	}
+
+	std::getline(header, number, ',');
+	gSaveNumber = std::stoi(number);
+
+	std::getline(header, number, ',');
+	demoGraphicsWidth = std::stoi(number);
+
+	std::getline(header, number, ',');
+	demoGraphicsHeight = std::stoi(number);
+
+	while (std::getline(demofile, line)) {
+		std::stringstream command(line);
+
+		std::getline(command, number, ',');
+		int typeNum = std::stoi(number);
+		auto type = static_cast<DemoMsgType>(typeNum);
+
+		std::getline(command, number, ',');
+		float progressToNextGameTick = std::stof(number);
+
+		switch (type) {
+		case DemoMsgType::Message: {
+			std::getline(command, number, ',');
+			uint32_t message = std::stoi(number);
+			std::getline(command, number, ',');
+			int32_t wParam = std::stoi(number);
+			std::getline(command, number, ',');
+			int32_t lParam = std::stoi(number);
+			PumpDemoMessage(type, message, wParam, lParam, progressToNextGameTick);
+			break;
+		}
+		default:
+			PumpDemoMessage(type, 0, 0, 0, progressToNextGameTick);
+			break;
+		}
+	}
+
+	demofile.close();
+
+	demoModeLastTick = SDL_GetTicks();
+
+	return true;
+}
+
+} // namespace
+
+namespace demo {
+
+int demoNumber = -1;
+bool timedemo = false;
+int recordNumber = -1;
+bool createDemoReference = false;
```

Comment:
Seems like these should be in the anonymous namespace above?

Reply:
They could cause they are not exported.
But I couldn't figure out how to write:
`demo::demoNumber = demoNumber;`
with a anonymous namespace, cause both the global variable and the parameter have the same name.

Alternative I could rename the function parameter (to `demoNumb` or something else) or the global variable (to `gDemoNumber` or something else).

Any preference? 🙂

Reply:
They are globals, style guide says globals start with a capital :)
```suggestion
int DemoNumber = -1;
bool Timedemo = false;
int RecordNumber = -1;
bool CreateDemoReference = false;
```

Reply:
Missed this one, as I often see globals in old name. Thank you, fixed. 🙂